### PR TITLE
Update testing to new approach fixes #42

### LINKED ISSUE #42: Update testing to new approach
Update for testing with cloud-build CI

### PR DESCRIPTION
**Pull request by ingwarr**
* Fixes https://github.com/terraform-google-modules/terraform-google-container-vm/issues/42

### CODE REVIEW

**Comment by ingwarr** [comment]
As I see the lint tests are broken because of ".terraform/* should be moved to ignore list. Also as I see managed_instance_group depends on  `http-lb`(updated to use with  0.12 terraform but not merged yet) and `mig` module(not updated to use with 0.12 terraform)

**Comment by ingwarr** [comment]
image
image

**Comment by kopachevsky** [comment]
I see tests are failing

File: examples/managed_instance_group/main.tf [hunk 1]
```diff
@@ -82,65 +72,66 @@ module "mig_template" {
...
-  subnetwork        = var.subnetwork
+  network    = google_compute_network.default.self_link
+  subnetwork = var.subnetwork
 }
-
 module "http-lb" {
-  source            = "../../../terraform-google-lb-http/"
-  project           = var.project_id
-  name              = "${var.mig_name}-lb"
+  source  = "github.com/terraform-google-modules/terraform-google-lb-http"
```
Comment (aaron-lane) [inline examples/managed_instance_group/main.tf]: This should be sourced from the Terraform Registry with a version.

```suggestion
  source  = "GoogleCloudPlatform/lb-http/google"
  version = "~> 1.0"
 
```

File: examples/simple_instance/variables.tf [hunk 2]
```diff
@@ -34,3 +34,12 @@ variable "instance_name" {
 variable "region" {
   description = "The GCP region to deploy instances into"
 }
+
+variable "client_email" {
```
Comment (aaron-lane) [inline examples/simple_instance/variables.tf]: All public variables should have descriptions.

File: .kitchen.yml [hunk 3]
```diff
@@ -62,6 +62,7 @@ suites:
         key_files:
           - ./test/fixtures/instance_with_attached_disk/ssh/key
         user: user
+        verify_host_key: false
```
[the remark this answers is not in the document]
Comment (aaron-lane) [reply .kitchen.yml]: Let's remove it if there is no justification.

File: modules/cos-coredns/assets/cloud-config.yaml [hunk 4]
```diff
@@ -1,5 +1,3 @@
-#cloud-config
```
Comment (aaron-lane) [inline modules/cos-coredns/assets/cloud-config.yaml]: @ludoo is there significance to this comment or can it be removed?

File: test/setup/main.tf [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  source  = "terraform-google-modules/project-factory/google"
+  version = "~> 3.0"
+
+  name              = "ci-container-vm"
+  random_project_id = true
+  org_id            = var.org_id
+  folder_id         = var.folder_id
+  billing_account   = var.billing_account
+
+  activate_apis = [
```
Comment (aaron-lane) [inline test/setup/main.tf]: This list seems excessive; the README only mentions `compute.googleapis.com`.

File: test/integration/instance_with_attached_disk/controls/docker.rb [hunk 6]
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+require "docker"
```
Comment (aaron-lane) [inline test/integration/instance_with_attached_disk/controls/docker.rb]: The `docker` method is provided by InSpec. What is the purpose of this import?

File: test/setup/main.tf (hunk 5 above)
Comment (aaron-lane) [reply test/setup/main.tf]: This list should be limited to APIs which are needed to run the tests. Any APIs required by the root module or submodules need to be documented in the appropriate READMEs.

File: modules/cos-coredns/assets/cloud-config.yaml (hunk 4 above)
Comment (ludoo) [reply modules/cos-coredns/assets/cloud-config.yaml]: it's actually needed. Linter should not break (license is still valid if it does not start on the first line afaik).

File: examples/simple_instance/variables.tf (hunk 2 above)
Comment (ingwarr) [reply examples/simple_instance/variables.tf]: Fixed

File: test/integration/instance_with_attached_disk/controls/docker.rb (hunk 6 above)
Comment (ingwarr) [reply test/integration/instance_with_attached_disk/controls/docker.rb]: Fixed

File: test/setup/main.tf (hunk 5 above)
Comment (ingwarr) [reply test/setup/main.tf]: Fixed

File: modules/cos-coredns/assets/cloud-config.yaml (hunk 4 above)
Comment (ingwarr) [reply modules/cos-coredns/assets/cloud-config.yaml]: removed

File: .dockerignore [hunk 7]
```diff
@@ -0,0 +1,10 @@
+.git
+.terraform
+.terraform.d
+.kitchen
+terraform.tfstate.d
+test/fixtures/*/.terraform
+test/fixtures/*/terraform.tfstate.d
+examples/.kitchen
```
[the remark this answers is not in the document]
Comment (ingwarr) [reply .dockerignore]: It copied from templates and not affected tests

**Comment by paulpalamarchuk** [comment]
This PR, also fixes #49

File: Makefile [hunk 8]
```diff
@@ -12,122 +12,74 @@
...
+# Enter docker container for local development
 .PHONY: docker_run
 docker_run:
 	docker run --rm -it \
-		-e CLOUDSDK_AUTH_CREDENTIAL_FILE_OVERRIDE=${CREDENTIALS_PATH} \
-		-e GOOGLE_APPLICATION_CREDENTIALS=${CREDENTIALS_PATH} \
-		-v "$(CURDIR)":/cft/workdir \
-		${DOCKER_REPO_BASE_KITCHEN_TERRAFORM} \
+		-e SERVICE_ACCOUNT_JSON \
+		-v $(CURDIR):/workspace \
```
Comment (aaron-lane) [inline Makefile]: Quotations around `$(CURDIR)` need to be maintained for compatibility with macOS.

Comment (paulpalamarchuk) [reply Makefile]: Added

File: modules/cos-coredns/assets/cloud-config.yaml (hunk 4 above)
Comment (paulpalamarchuk) [reply modules/cos-coredns/assets/cloud-config.yaml]: Comments restored. Also excluded `cloud-config.yaml`-files from [check_headers func](https://github.com/terraform-google-modules/terraform-google-container-vm/pull/43/files#diff-9aca4f6ccbe56e14f9edf295b7db1d3f)

**Comment by paulpalamarchuk** [comment]
@aaron-lane at the moment example `managed_instance_group` failing with:
```
$ terraform plan

Error: Invalid value for module argument

  on main.tf line 119, in module "http-lb":
 119:   backends = {
 120:     "0" = [
 121:       {
 122:         group = module.mig.instance_group
 123:       },
 124:     ]
 125:   }

The given value is not suitable for child module variable "backends" defined
at ../../../terraform-google-lb-http/variables.tf:56,1-20: element "0":
element 0: attributes "balancing_mode", "capacity_scaler", "description",
"max_connections", "max_connections_per_instance", "max_rate",
"max_rate_per_instance", and "max_utilization" are required.

```
But on the current branch, in seems ok:
image

There were a lot of changes, in that PR. Also master branch changed meanwhile. I think image name, already fixed in master, but got another error. 
It's hard to find what exactly fixed that issue

**Pull request merged**